File: 2024/19/solution.rb
f = File.open(__dir__ + "/input.txt")
towels = f.gets.split(',').map(&:strip)
f.gets
patterns = f.readlines.map { |line| line.strip }

def find(pattern, towels, memo = {})
  return 1 if pattern.empty?
  memo[pattern] ||= towels.map { |towel| pattern.start_with?(towel) ? find(pattern[towel.size..], towels, memo) : nil }.compact.sum
end

memo = {}
puts "solution1: " + patterns.map { |pattern| find(pattern, towels, memo) }.compact.select { |x| x > 0 }.count.to_s
puts "solution2: " + patterns.map { |pattern| find(pattern, towels, memo) }.compact.sum.to_s

File: 2024/19/solution.jl
using Memoization

@memoize function find(pattern, towels)
    if isempty(pattern)
        1
    else
        sum(Int[find(pattern[length(towel)+1:end], towels) for towel in towels if startswith(pattern, towel)])
    end
end

f = open(joinpath(@__DIR__, "input.txt"), "r")
towels = split(readline(f), ", ")
readline(f)
patterns = readlines(f)
close(f)

function process((solution1, solution2), pattern)
    found = find(pattern, towels)
    if found > 0
        solution1 += 1
    end
    (solution1, solution2 + found)
end

@time solution1, solution2 = reduce(process, patterns, init=(0, 0))
println("solution1: ", solution1)
println("solution2: ", solution2)
